lib/event.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2018, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <metal/assert.h>
+#include <metal/event.h>
+#include <metal/list.h>
+
+static METAL_DECLARE_LIST(metal_default_event_queue);
+
+void metal_event_uninit(struct metal_event *event)
+{
+	metal_assert(event != NULL);
+	metal_spinlock_acquire(&event->slock);
+	_metal_event_pop(event);
+	event->state = METAL_EVENT_IDLE;
+	event->hd.func = NULL;
+	event->hd.arg = NULL;
+	metal_spinlock_release(&event->slock);
+}
+
+int metal_event_post(struct metal_event *event)
+{
+	int ret;
+
+	metal_assert(event != NULL);
+	metal_spinlock_acquire(&event->slock);
+	if (event->state & METAL_EVENT_ENQUEUED) {
+		return 0;
+	} else if (event->state & METAL_EVENT_ENABLED) {
+		ret = _metal_event_post(event);
+		if (ret == 0)
+			event->state |= METAL_EVENT_ENQUEUED;
+	} else {
+		ret = 0;
+		event->state |= METAL_EVENT_PENDING;
+	}
+	metal_spinlock_release(&event->slock);
+	return ret;
+}
+
+void metal_event_pop(struct metal_event *event)
+{
+	metal_assert(event != NULL);
+	metal_spinlock_acquire(&event->slock);
+	_metal_event_pop(event);
+	event->state &= ~METAL_EVENT_ENQUEUED;
+	metal_spinlock_release(&event->slock);
+}
+
+int metal_event_register_handler(struct metal_event *event,
+				 metal_event_hd_func func,
+				 void *arg)
+{
+	metal_assert(event != NULL);
+	event->hd.func = func;
+	event->hd.arg = arg;
+	return 0;
+}
+
+void metal_event_enable(struct metal_event *event)
+{
+	metal_assert(event != NULL);
+	metal_spinlock_acquire(&event->slock);
+	event->state |= METAL_EVENT_ENABLED;
+	if (event->state & METAL_EVENT_PENDING)
+		event->state &= ~METAL_EVENT_PENDING;
+	metal_spinlock_release(&event->slock);
+	metal_event_post(event);
+}
+
+void metal_event_disable(struct metal_event *event)
+{
```

Comment:
does disable will discard the events? E.g. if there is an interrupt of inter-processor communication raised while the inter-processor communication event is disabled, shall the event marked as PENDING or ignore it?
How about have EVENT_HANDLER_ENABLED state?

Reply:
Would be useful to be able to discard any pending and queued events.